### PR TITLE
Some optimizations w.r.t. shared parameters for module level nonrecursive policy

### PR DESCRIPTION
**Pull request by LinjianMa**

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -3175,7 +3187,12 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 and self._param_exec_order_state == ParamExecOrderState.UNINITIALIZED
             ):
                 # In self._handles_exec_order, the handles are ordered based on
-                # the execution order in the backward pass in the first iteration.
+                # the gradient ready order in the backward pass in the first
+                # iteration.
+                # Note: the gradient ready order of `FlatParameter` may not be
+                # consistent with the order of `nn.Parameter` in the original
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: Maybe (to clarify it is not referring to the order in which the parameters are registered/defined):
```suggestion
                # consistent with the execution order of `nn.Parameter` in the original
```

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 2]
```diff
@@ -72,45 +72,62 @@
...
-    def __init__(self) -> None:
+    def __init__(self, embed) -> None:
         super().__init__()
+        self.embed = embed
         self.conv1 = nn.Conv2d(3, 6, 5)
         self.pool = nn.MaxPool2d(2, 2)
         self.conv2 = nn.Conv2d(6, 16, 5)
-        # Handles the edge case when this parameter is wrapped together with
+        self.fc = nn.Linear(16 * 5 * 5, 8)
+        # Handles the edge chase when this parameter is wrapped together with
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]:
```suggestion
        # Handles the edge case when this parameter is wrapped together with
```

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 3]
```diff
@@ -154,7 +171,7 @@ def _check_fsdp_param_parity(self, fsdp_model, ref_model):
                 print(f"[Rank {self.rank}] fsdp names: {[n for n, _ in fsdp_model.named_parameters()]}")
             self.assertEqual(len(fsdp_named_params), len(ref_named_params))
             for (_, p1), (_, p2) in zip(fsdp_named_params, ref_named_params):
-                torch.testing.assert_close(p1, p2, rtol=1e-4, atol=1e-4)
+                torch.testing.assert_close(p1, p2, rtol=1e-2, atol=1e-3)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: This is kind of concerning. For me historically, sometimes when the error is at this magnitude, it was an actual error. We can discuss offline how we might address this.

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 4]
```diff
@@ -72,45 +72,62 @@
...
 
 class CNN(nn.Module):
     def __init__(self) -> None:
         super().__init__()
-        self.conv = DoubleConv()
-        self.fc1 = nn.Linear(16 * 5 * 5, 20)
-        self.fc2 = nn.Linear(20, 20)
-        self.fc3 = nn.Linear(20, 20)
-        self.root_weight = torch.nn.Parameter(torch.randn(20, 20))
+        # test shared module
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py:112]: Is there any notable difference to FSDP about a shared parameter from a shared module (like here) versus a shared parameter that is directly shared across two different modules (like in `TransformerWithSharedParams`)?

Curious: Have you seen the latter case in practice?

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 5]
```diff
@@ -291,6 +308,38 @@ def test_train_with_full_state_dict(
             fsdp_model, fsdp_optim, ddp_model, ddp_optim, inp_shape, num_iters,
         )
 
+    @skip_if_lt_x_gpu(2)
+    def test_wrap_structure(self):
+        """
+        Tests that ``nn.Parameter`` are corrected grouped as ``FlatParameter``.
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]:
```suggestion
        Tests that ``nn.Parameter`` are correctly grouped as ``FlatParameter``.
```

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 6]
```diff
@@ -291,6 +308,38 @@ def test_train_with_full_state_dict(
...
+            {'root_weight', 'shared.weight'},
+            {'fc3.weight', 'fc3.bias'},
+            {'fc2.weight', 'fc2.bias'},
+            {'conv2.weight', 'conv2.conv1.weight', 'conv2.conv1.bias', 'conv2.conv2.weight', 'conv2.conv2.bias'},
+            {'conv2.fc.weight', 'conv2.fc.bias'},
+            {'conv1.weight', 'conv1.conv1.weight', 'conv1.conv1.bias', 'conv1.conv2.weight', 'conv1.conv2.bias'},
+            {'conv1.fc.weight', 'conv1.fc.bias'},
+        ]
+        for handle in fsdp_model._handles:
+            self.assertTrue(
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: nit: can use `self.assertIn()`

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 7]
```diff
@@ -291,6 +308,38 @@ def test_train_with_full_state_dict(
...
+        """
+        model = CNN().cuda()
+        group = dist.distributed_c10d._get_default_group()
+        # Used to test non-default module_level_group_policy
+        module_level_policy = functools.partial(
+            transformer_auto_wrap_policy,
+            transformer_layer_cls={DoubleConv, nn.Linear},
+        )
+        param_exec_order_policy = ParamExecOrderPolicy(
+                handle_init_mode=HandleInitMode.MODULE_LEVEL,
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: I think the indent is off here. You can check with `lintrunner`.

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 8]
```diff
@@ -291,6 +308,38 @@ def test_train_with_full_state_dict(
...
+            {'fc2.weight', 'fc2.bias'},
+            {'conv2.weight', 'conv2.conv1.weight', 'conv2.conv1.bias', 'conv2.conv2.weight', 'conv2.conv2.bias'},
+            {'conv2.fc.weight', 'conv2.fc.bias'},
+            {'conv1.weight', 'conv1.conv1.weight', 'conv1.conv1.bias', 'conv1.conv2.weight', 'conv1.conv2.bias'},
+            {'conv1.fc.weight', 'conv1.fc.bias'},
+        ]
+        for handle in fsdp_model._handles:
+            self.assertTrue(
+                set(handle.flat_param._prefixed_param_names) in correct_prefixed_param_names
+            )
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py:344]: Should we add a `self.assertEqual(len(fsdp_model._handles), len(correct_prefixed_param_names))` to ensure that the former is not a strict subset of the latter?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 9]
```diff
@@ -3175,7 +3187,12 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
...
+                # model, since for each `FlatParameter`, its handle will always
+                # run `_unflatten` in `_pre_forward` before the module forward.
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: Sorry that you already explained this to me a few times, but I do not feel like I fully understand why this is happening. Let me ask you about it directly.

IIUC, in particular it is the `view()` operator in `_unflatten()` that induces this effect on the gradient ready order. However, I forget _why_ this is the case.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 10]
```diff
@@ -991,12 +982,33 @@ def _get_params_per_wrapped_module(self, root_module) -> List[List[nn.Parameter]
...
+        traversed_params: Set[nn.Parameter] = set()
+        # Modules in `modules_to_wrap` are ordered based on post-order traversal.
+        # We record parameters based on reverse postordering, which is a
+        # topological sort so each shared parameter is guaranteed to be grouped
+        # with its lowest common ancester module's parameters.
+        modules_to_wrap.reverse()
+        for module_to_wrap in modules_to_wrap:
+            # Perform a BFS starting from module, and record all untraversed
+            # parameters that are not the parameter of any other module in
+            # modules_to_wrap.
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: nit: 
```suggestion
            # `modules_to_wrap`.
```

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 11]
```diff
@@ -991,12 +982,33 @@ def _get_params_per_wrapped_module(self, root_module) -> List[List[nn.Parameter]
...
-            p for p in root_module.parameters() if p not in traversed_param_set
-        ]
-        params_per_wrapped_module.append(remaining_params)
-        return params_per_wrapped_module
+        if root_module not in modules_to_wrap:
+            modules_to_wrap.append(root_module)
+
+        params_per_wrap: List[List[nn.Parameter]] = []
+        traversed_params: Set[nn.Parameter] = set()
+        # Modules in `modules_to_wrap` are ordered based on post-order traversal.
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]: nit: 
```suggestion
        # Constructing `modules_to_wrap` with `_recursive_wrap` orders the modules based on a post-order traversal.
```
I just wanted to clarify that this post-order traversal comes from `_recursive_wrap`.

That being said, maybe we should mention in the `_recursive_wrap` docstring that it is post-order if not already since now we are relying on that.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 12]
```diff
@@ -991,12 +982,33 @@ def _get_params_per_wrapped_module(self, root_module) -> List[List[nn.Parameter]
...
+        # We record parameters based on reverse postordering, which is a
+        # topological sort so each shared parameter is guaranteed to be grouped
+        # with its lowest common ancester module's parameters.
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]:
```suggestion
        # with its lowest common ancestor module's parameters.
```

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 13]
```diff
@@ -991,12 +982,33 @@ def _get_params_per_wrapped_module(self, root_module) -> List[List[nn.Parameter]
...
+        modules_to_wrap.reverse()
+        for module_to_wrap in modules_to_wrap:
+            # Perform a BFS starting from module, and record all untraversed
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py]:
```suggestion
            # Perform a BFS starting from `module_to_wrap`, and record all untraversed
```
or
```suggestion
            # Perform a BFS starting from the module, and record all untraversed
```

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py [hunk 14]
```diff
@@ -291,6 +308,38 @@ def test_train_with_full_state_dict(
...
+            transformer_auto_wrap_policy,
+            transformer_layer_cls={DoubleConv, nn.Linear},
+        )
+        param_exec_order_policy = ParamExecOrderPolicy(
+                handle_init_mode=HandleInitMode.MODULE_LEVEL,
+                module_level_group_policy=module_level_policy,
+            )
+        fsdp_model = FSDP(model, group, auto_wrap_policy=param_exec_order_policy)
+        correct_prefixed_param_names = [
+            # here `shared.weight` should be in the same wrap as `root_weight`
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_param_exec_order_policy.py:329]: Add the explanation for why? Something like:
```suggestion
            # here `shared.weight` should be in the same wrap as `root_weight`
            # because the root module `model` is the lowest common ancestor of
            # `the modules conv1` and `conv2` that share `shared.weight`
```
(can rewrap/reformat as needed)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 15]
```diff
@@ -991,12 +982,33 @@ def _get_params_per_wrapped_module(self, root_module) -> List[List[nn.Parameter]
             ignored_params=[],
             only_wrap_children=False,
             )
-        # parameters not wrapped in ``policy`` are grouped together.
-        remaining_params = [
-            p for p in root_module.parameters() if p not in traversed_param_set
-        ]
-        params_per_wrapped_module.append(remaining_params)
-        return params_per_wrapped_module
+        if root_module not in modules_to_wrap:
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:986]: When would this conditional be true or and when would it be false?

**Review by awgu — COMMENTED** [review]
Awesome work! This flattening the shared parameters with the LCA module's parameters is for the first iteration right. It just affects the initial wraps?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 11 above)
Comment (LinjianMa) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: not relevant to this PR, but I think it's better to change the name of `_recursive_wrap` to something like `_wrap_post_order_traversal`. It's not clear what `_recursive_wrap` means

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 15 above)
Comment (LinjianMa) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:986]: This is used when `root_module` is not wrapped in `_recursive_wrap`. For example, `transformer_auto_wrap_policy` usually won't wrap `root_module`. I will add one comment.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 9 above)
Comment (LinjianMa) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: synced offline

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py (hunk 3 above)
Comment (LinjianMa) [reply test/distributed/fsdp/test_fsdp_param_exec_order_policy.py]: I changed the parameters back. But this test may fail sometime. We can think of how to fix this later

File: test/distributed/fsdp/test_fsdp_param_exec_order_policy.py (hunk 4 above)
Comment (LinjianMa) [reply test/distributed/fsdp/test_fsdp_param_exec_order_policy.py:112]: I haven't seen the latter case yet, but I think that makes sense and we should be able to handle that. We can add test cases for that later

**Comment by LinjianMa** [comment]
> Awesome work! This flattening the shared parameters with the LCA module's parameters is for the first iteration right. It just affects the initial wraps?

Yes! It just affects the initial wrap

**Review by awgu — APPROVED** [review]
LGTM! I will merge.
